spaghetto/src/Spago/Command/Ls.purs

```diff
@@ -1,3 +1,120 @@
+import Spago.Config as Config
+import Spago.Json (stringifyJson)
+
+data JsonFlag = JsonOutputNo | JsonOutputYes
+
+type LsEnv =
+  { dependencies :: PackageSet
+  , logOptions :: LogOptions
+  , workspace :: Workspace
+  }
+
+type JsonPackageOutput =
+  { packageName :: String
+  -- TODO: Was Json Value, has tag
+  -- See https://github.com/purescript/spago/blob/ec22308d181340382494e5bd6ef9102a3c87c0bb/src/Spago/Types.hs#L38-L47
+  , repo :: String
+  , version :: String
+  }
+
+jsonPackageOutputCodec :: JsonCodec JsonPackageOutput
+jsonPackageOutputCodec = CAR.object "JsonPackageOutput"
+  { packageName: CA.string
+  , repo: CA.string
+  , version: CA.string
+  }
+
+listPackageSet :: JsonFlag -> Spago LsEnv Unit
+listPackageSet jsonFlag = do
+  logDebug "Running `listPackageSet`"
+  { workspace } <- ask
+  output $ formatPackageNames jsonFlag (Map.toUnfoldable workspace.packageSet)
+
+listPackages :: IncludeTransitive -> JsonFlag -> Spago LsEnv Unit
+listPackages packagesFilter jsonFlag = do
+  logDebug "Running `listPackages`"
+  { dependencies, workspace } <- ask
+  let
+    packagesToList = Map.toUnfoldable case packagesFilter of
+      IncludeTransitive -> dependencies
+      NoIncludeTransitive ->
+        let
+          workspacePackages = Config.getWorkspacePackages workspace.packageSet
+          toDependencyNames (Dependencies deps) = Map.keys deps
+          directDependencies = unions $ ((toDependencyNames <<< _.dependencies <<< _.package) <$> workspacePackages)
+        in
+          filterKeys (_ `elem` directDependencies) dependencies
+  case packagesToList of
+    [] -> logWarn "There are no dependencies listed in your spago.dhall"
+    _ -> output $ formatPackageNames jsonFlag packagesToList
+
+formatPackageNames :: forall a. JsonFlag -> Array (PackageName /\ Package) -> OutputFormat a
+formatPackageNames = case _ of
+  JsonOutputYes -> OutputLines <<< formatPackageNamesJson
+  JsonOutputNo -> OutputLines <<< formatPackageNamesText
+  where
+  formatPackageNamesJson :: Array (PackageName /\ Package) -> Array String
+  formatPackageNamesJson pkgs = (stringifyJson jsonPackageOutputCodec <<< asJson <$> pkgs)
+    where
+    asJson (name /\ package) =
+      { packageName: PackageName.print name
+      -- TODO: Do we want to stay backward compatible and have a tagged remote and local type?
+      -- Or do we create a codec for package (with 4 variants)?
+      , repo: "TODO"
+      , version: showVersion package
+      }
+
+  -- | Format all the package names from the configuration
+  formatPackageNamesText :: Array (PackageName /\ Package) -> Array String
+  formatPackageNamesText pkgs =
+    let
+
+      -- TODO: Currently prints git/remote packages as local packages
+      showLocation :: PackageName -> Package -> String
+      showLocation _ (GitPackage gitPackage) = "Remote " <> surroundQuote gitPackage.git
+      showLocation packageName package = "Local " <> surroundQuote (getPackageLocation packageName package)
```

Comment:
:bug: Here the behavior is definitely off right now. 

> Local ".spago/packages/strings-6.0.1

Printing `Local` but what I want is the git package. See other comment.

Reply:
Yeah, this is because `getPackageLocation` is not we need for this, since it's for getting the path of the downloaded package on disk.

I guess the textual representation needs a little more thought than the JSON version (since for that one it's enough to add a tag to denote the kind of package, and then just dump the representation of each package in JSON).

The current format includes package name, version and location.
We could keep the same format and have the different kind of packages map to that sensibly:
| Package type  | Version | Location |
| ------------- | ------- | -------- |
| registry      | version | ""       |
| git           | ref     | repo     |
| local         | "local" | path     |
| workspace     | "workspace" | path |

Reply:
Tried the table (based on dodo-printer) in 415149d6958a929a22a95ea7b85a1d1545e4b04a

Reply:
See https://github.com/purescript/spago/pull/947#discussion_r1111105804